images/vm-route-forge/internal/controller/route/route_controller.go

```diff
@@ -0,0 +1,292 @@
+		return vmInformer.Informer().HasSynced() && nodeInformer.Informer().HasSynced()
+	}
+
+	return routeController, nil
+}
+
+type Controller struct {
+	vmIndexer      cache.Indexer
+	vmLister       virtlisters.VirtualMachineLister
+	nodeLister     corelisters.NodeLister
+	hostReconciler Runnable
+	hasSynced      cache.InformerSynced
+	queue          workqueue.RateLimitingInterface
+	cache          cache2.Cache
+	netlinkMgr     *netlinkmanager.Manager
+	log            *logr.Logger
+}
+
+func (c *Controller) addVirtualMachine(obj interface{}) {
+	vm, ok := obj.(*v1alpha2.VirtualMachine)
+	if !ok {
+		return
+	}
+	c.enqueueVirtualMachine(vm)
+}
+
+func (c *Controller) deleteVirtualMachine(obj interface{}) {
+	vm, ok := obj.(*v1alpha2.VirtualMachine)
+	if !ok {
+		return
+	}
+	c.enqueueVirtualMachine(vm)
+}
+func (c *Controller) updateVirtualMachine(oldObj interface{}, newObj interface{}) {
+	oldVm, ok := oldObj.(*v1alpha2.VirtualMachine)
+	if !ok {
+		return
+	}
+	newVm, ok := newObj.(*v1alpha2.VirtualMachine)
+	if !ok {
+		return
+	}
+	if oldVm.Status.IPAddress != newVm.Status.IPAddress || oldVm.Status.Node != newVm.Status.Node {
+		c.enqueueVirtualMachine(newVm)
+	}
+}
+
+func (c *Controller) addNode(_ interface{}) {
+	// Do nothing
+}
+
+func (c *Controller) deleteNode(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		return
+	}
+	vms, err := c.vmLister.List(labels.Everything())
+	if err != nil {
+		c.log.Error(err, "failed to list virtual machines")
+		return
+	}
+
+	for _, vm := range vms {
+		if vm.Status.Node == node.Name {
+			c.enqueueVirtualMachine(vm)
+		}
+	}
+
+}
+func (c *Controller) updateNode(oldObj interface{}, newObj interface{}) {
+	oldNode, ok := oldObj.(*corev1.Node)
+	if !ok {
+		return
+	}
+	newNode, ok := newObj.(*corev1.Node)
+	if !ok {
+		return
+	}
+	var oldIP, newIP string
+	for _, ip := range oldNode.Status.Addresses {
+		if ip.Type == corev1.NodeInternalIP {
+			oldIP = ip.Address
+			break
+		}
+	}
+	for _, ip := range newNode.Status.Addresses {
+		if ip.Type == corev1.NodeInternalIP {
+			newIP = ip.Address
+			break
+		}
+	}
+	if oldIP == newIP {
+		return
+	}
+	vms, err := c.vmLister.List(labels.Everything())
+	if err != nil {
+		c.log.Error(err, "failed to list virtual machines")
+		return
+	}
+
+	for _, vm := range vms {
+		if vm.Status.Node == newNode.Name {
+			c.enqueueVirtualMachine(vm)
+		}
+	}
+}
+
+func (c *Controller) enqueueVirtualMachine(vm *v1alpha2.VirtualMachine) {
+	key, err := KeyFunc(vm)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", vm, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.log.Info("Starting namespace controller")
+	defer c.log.Info("Shutting down namespace controller")
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.hasSynced) {
+		return
+	}
```

Comment:
Isn't it a disaster? Should we log an error at this point?

Reply:
maybe, but I think it's not necessary because we need to wait for cache synchronization before running handlers. If the cache is not synchronized, the controller will stop. For example, in the namespace controller it works the same way [(link)](https://github.com/kubernetes/kubernetes/blob/0fc167103189a4d462c3cc7a17b360c3d998f4bf/pkg/controller/namespace/namespace_controller.go#L205). But i added log message